game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java

```diff
@@ -0,0 +1,125 @@
+package games.strategy.triplea.delegate.battle.steps.fire;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.delegate.Matches;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import lombok.Value;
+import org.triplea.java.collections.CollectionUtils;
+
+/**
+ * Group of units that are firing on targets
+ *
+ * <p>If the group is suicideOnHit, then all of the units should have the same unit type
+ */
+@Value
+public class FiringGroup {
+
+  private static final UnitType NON_SUICIDE_MULTIMAP_KEY = new UnitType("nonsuicide", null);
+
+  String displayName;
+  String groupName;
+  Collection<Unit> firingUnits;
+  Collection<Unit> targetUnits;
+  boolean suicideOnHit;
+
+  private FiringGroup(
+      final String displayName,
+      final String groupName,
+      final Collection<Unit> firingUnits,
+      final Collection<Unit> targetUnits) {
+    this.displayName = displayName;
+    this.groupName = groupName;
+    this.firingUnits = firingUnits;
+    this.targetUnits = targetUnits;
+    this.suicideOnHit = this.firingUnits.stream().allMatch(Matches.unitIsSuicideOnHit());
+  }
+
+  public Collection<Unit> getTargetUnits() {
+    return Collections.unmodifiableCollection(targetUnits);
```

Comment:
I wonder if callers could just avoid mutating the return value and/or copy as needed. If we can skip the defensive copy, we can have a performance improvement and simpler code in this class as a simpler getter would do.

Reply:
Something to hope for :)

Reply:
The copies can add up to be significant. Can we do better than hoping? The java8 streaming API is particularly helpful for operating on returned collections as the collection is not mutated.

Reply:
The return value of this method is passed off to other methods and those pass off to other methods.  So the code path isn't easy to follow.

But I can't remember why exactly I added unmodifiableCollection.  I thought it was because that is how the original code was but I can't find any evidence of that now.  So, I've removed the unmodifiableCollection part.